src/main/java/com/hms_networks/sc/canary/api/CanaryDataPosterThread.java

```diff
@@ -39,14 +40,28 @@ public void run() {
         Logger.LOG_DEBUG("Sending completed payload to Canary");
         CanaryApiResponseStatus requestStatus = CanaryApiRequestSender.processRequest(request);
 
-        if (requestStatus != CanaryApiResponseStatus.GOOD_REQUEST) {
-          Logger.LOG_WARN("Unable to send payload to Canary");
-        } else {
+        if (requestStatus == CanaryApiResponseStatus.GOOD_REQUEST) {
           SessionManager.updateTokenExpiration();
           boolean removed = CanaryDataPayloadManager.removeNextPayload();
           if (!removed) {
             Logger.LOG_WARN("Unable to remove payload from queue");
           }
+        } else if (requestStatus == CanaryApiResponseStatus.ERROR_WAIT_FOR_EXPIRE) {
+          Logger.LOG_WARN(
+              "Waiting for existing sessions to expire before sending more data to Canary");
+          try {
+            final long apiClientTimeoutMillis =
+                SCTimeUnit.SECONDS.toMillis(
+                    CanaryConnectorMain.getConnectorConfig().getApiClientTimeoutSeconds());
+            CanaryConnectorMain.getInstance().setDataPollingBlocked(true);
+            Thread.sleep(apiClientTimeoutMillis);
+            CanaryConnectorMain.getInstance().setDataPollingBlocked(false);
+          } catch (InterruptedException e) {
+            Logger.LOG_SERIOUS("An error occurred while waiting for existing sessions to expire.");
+            Logger.LOG_EXCEPTION(e);
+          }
+        } else {
+          Logger.LOG_WARN("Unable to send payload to Canary");

```

Comment:
should you log the request status here?

Reply:
No, as it would be redundant. The sender class is the central point for HTTP(s) requests/responses and takes care of this. 
![image](https://github.com/hms-networks/flexy-canary-connector/assets/51409174/ee4b2f00-f643-46f0-88cd-45ac99721b9f)
![image](https://github.com/hms-networks/flexy-canary-connector/assets/51409174/80d81cc2-b7e5-4382-9cac-e3cdb27bad94)